File: apps/web/app/lib/walletAnalysis.ts
"use client";

import { fetchApiJson } from "./api";

export type WalletAnalysisMetricStatus = "estimated" | "unavailable" | "verified";
export type WalletAnalysisConfidence = "high" | "low" | "medium";
export type WalletAnalysisSignalStrength = "strong" | "moderate" | "weak";
export type WalletAnalysisJobStatus =
  | "pending"
  | "resolving_market"
  | "discovering_wallets"
  | "analyzing_wallets"
  | "scoring"
  | "completed"
  | "partial"
  | "failed"
  | "cancelled";
export type WalletAnalysisRunState = "progressed" | "already_running" | "no_work_remaining" | "failed";
export type WalletAnalysisCandidateSortBy = "created_at" | "pnl_30d" | "score" | "volume_30d" | "win_rate_30d";
export type WalletAnalysisSortOrder = "asc" | "desc";
export type WalletProfileStatus = "candidate" | "watching" | "demo_follow" | "paused" | "rejected";
export type MarketSignalStatus =
  | "pending_resolution"
  | "resolved_hit"
  | "resolved_miss"
  | "cancelled"
  | "unknown"
  | "no_clear_signal";
export type MarketResolutionStatus = "open" | "resolved" | "cancelled" | "unknown";

export type WalletAnalysisOutcome = {
  label: string;
  side: string;
  token_id?: string | null;
};

export type WalletAnalysisJobProgress = {
  wallets_found: number;
  wallets_analyzed: number;
  wallets_with_sufficient_history: number;
  yes_wallets: number;
  no_wallets: number;
  current_batch: number;
};

export type WalletAnalysisSignalSummary = {
  id: string;
  predicted_side?: string | null;
  predicted_outcome?: string | null;
  polysignal_score?: string | number | null;
  confidence: WalletAnalysisConfidence;
  data_confidence?: WalletAnalysisConfidence | null;
  signal_strength?: WalletAnalysisSignalStrength | null;
  signal_margin?: string | number | null;
  yes_score?: string | number | null;
  no_score?: string | number | null;
  outcome_scores_json?: Record<string, unknown> | null;
  outcome_wallet_counts_json?: Record<string, number> | null;
  signal_status: MarketSignalStatus;
  warnings_json: string[];
};

export type WalletAnalysisJobRead = {
  id: string;
  source_url: string;
  normalized_url: string;
  market_slug?: string | null;
  event_slug?: string | null;
  condition_id?: string | null;
  market_title?: string | null;
  status: WalletAnalysisJobStatus;
  outcomes: WalletAnalysisOutcome[];
  token_ids: string[];
  progress: WalletAnalysisJobProgress;
  result_json?: Record<string, unknown> | null;
  warnings: string[];
  status_detail?: string | null;
  error_message?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
  updated_at: string;
  candidates_count: number;
  signal_summary?: WalletAnalysisSignalSummary | null;
};

export type WalletAnalysisJobCreateResponse = {
  job_id: string;
  status: WalletAnalysisJobStatus;
  message: string;
  market: WalletAnalysisJobRead;
};

export type WalletAnalysisRunResponse = {
  job_id: string;
  status: WalletAnalysisJobStatus;
  run_state: WalletAnalysisRunState;
  message: string;
  wallets_found: number;
  wallets_analyzed: number;
  wallets_with_sufficient_history: number;
  candidates_count: number;
  warnings: string[];
  status_detail?: string | null;
  has_more: boolean;
  next_action?: string | null;
  signal_id?: string | null;
  signal_status?: string | null;
  market: WalletAnalysisJobRead;
};

export type WalletAnalysisCandidate = {
  id: string;
  job_id: string;
  wallet_address: string;
  outcome?: string | null;
  side?: string | null;
  token_id?: string | null;
  observed_market_position_usd?: string | number | null;
  score?: string | number | null;
  confidence: WalletAnalysisConfidence;
  roi_30d_status: WalletAnalysisMetricStatus;
  roi_30d_value?: string | number | null;
  win_rate_30d_status: WalletAnalysisMetricStatus;
  win_rate_30d_value?: string | number | null;
  pnl_30d_status: WalletAnalysisMetricStatus;
  pnl_30d_value?: string | number | null;
  trades_30d?: number | null;
  volume_30d?: string | number | null;
  markets_traded_30d?: number | null;
  last_activity_at?: string | null;
  reasons_json: string[];
  risks_json: string[];
  raw_summary_json?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type WalletAnalysisCandidateList = {
  items: WalletAnalysisCandidate[];
  total: number;
};

export type WalletProfileRead = {
  id: string;
  wallet_address: string;
  alias?: string | null;
  status: WalletProfileStatus;
  score?: string | number | null;
  confidence: WalletAnalysisConfidence;
  roi_30d_status: WalletAnalysisMetricStatus;
  roi_30d_value?: string | number | null;
  win_rate_30d_status: WalletAnalysisMetricStatus;
  win_rate_30d_value?: string | number | null;
  pnl_30d_status: WalletAnalysisMetricStatus;
  pnl_30d_value?: string | number | null;
  trades_30d?: number | null;
  volume_30d?: string | number | null;
  drawdown_30d_status: WalletAnalysisMetricStatus;
  drawdown_30d_value?: string | number | null;
  markets_traded_30d?: number | null;
  last_activity_at?: string | null;
  discovered_from_market?: string | null;
  discovered_from_url?: string | null;
  discovered_at?: string | null;
  reasons_json: string[];
  risks_json: string[];
  notes?: string | null;
  created_at: string;
  updated_at: string;
};

export type WalletProfileList = {
  items: WalletProfileRead[];
  total: number;
};

export type WalletProfileDemoFollowResponse = {
  profile: WalletProfileRead;
  copy_wallet: Record<string, unknown>;
  already_following: boolean;
  baseline_created_at: string;
  message: string;
};

export type PolySignalMarketSignal = {
  id: string;
  job_id?: string | null;
  source_url?: string | null;
  market_slug?: string | null;
  event_slug?: string | null;
  condition_id?: string | null;
  market_title?: string | null;
  predicted_side?: string | null;
  predicted_outcome?: string | null;
  polysignal_score?: string | number | null;
  confidence: WalletAnalysisConfidence;
  data_confidence?: WalletAnalysisConfidence | null;
  signal_strength?: WalletAnalysisSignalStrength | null;
  signal_margin?: string | number | null;
  yes_score?: string | number | null;
  no_score?: string | number | null;
  outcome_scores_json?: Record<string, unknown> | null;
  outcome_wallet_counts_json?: Record<string, number> | null;
  wallets_analyzed?: number | null;
  wallets_with_sufficient_history?: number | null;
  warnings_json: string[];
  signal_status: MarketSignalStatus;
  final_outcome?: string | null;
  final_resolution_source?: string | null;
  resolved_at?: string | null;
  top_wallets_json?: Array<Record<string, unknown>>;
  created_at: string;
  updated_at: string;
};

export type PolySignalMarketSignalMetricsBucket = {
  total: number;
  resolved_hit: number;
  resolved_miss: number;
  win_rate?: string | number | null;
};

export type PolySignalMarketSignalMetrics = {
  total: number;
  pending_resolution: number;
  resolved_hit: number;
  resolved_miss: number;
  cancelled: number;
  unknown: number;
  no_clear_signal: number;
  win_rate?: string | number | null;
  avg_score_resolved_hit?: string | number | null;
  avg_score_resolved_miss?: string | number | null;
  by_confidence: Record<string, PolySignalMarketSignalMetricsBucket>;
};

export type PolySignalMarketSignalList = {
  items: PolySignalMarketSignal[];
  total: number;
  metrics: PolySignalMarketSignalMetrics;
};

export type PolySignalMarketResolution = {
  status: MarketResolutionStatus;
  final_outcome?: string | null;
  source: string;
  confidence: WalletAnalysisConfidence;
  reason: string;
  checked_at: string;
};

export type PolySignalMarketSignalSettlement = {
  signal: PolySignalMarketSignal;
  resolution: PolySignalMarketResolution;
  changed: boolean;
};

export type PolySignalMarketSignalSettlePendingResponse = {
  checked: number;
  still_pending: number;
  resolved_hit: number;
  resolved_miss: number;
  cancelled: number;
  unknown: number;
  errors: number;
  items: PolySignalMarketSignalSettlement[];
};

export async function createWalletAnalysisJob(polymarketUrl: string): Promise<WalletAnalysisJobCreateResponse> {
  return fetchApiJson<WalletAnalysisJobCreateResponse>("/wallet-analysis/jobs", {
    body: JSON.stringify({ polymarket_url: polymarketUrl }),
    method: "POST",
  });
}

export async function runWalletAnalysisJobOnce(input: {
  batchSize?: number;
  historyLimit?: number;
  jobId: string;
  maxWallets?: number;
  maxWalletsDiscovery?: number;
  maxRuntimeSeconds?: number;
}): Promise<WalletAnalysisRunResponse> {
  return fetchApiJson<WalletAnalysisRunResponse>(`/wallet-analysis/jobs/${encodeURIComponent(input.jobId)}/run-once`, {
    body: JSON.stringify({
      batch_size: input.batchSize ?? 20,
      history_limit: input.historyLimit ?? 100,
      max_wallets: input.maxWallets ?? 50,
      max_wallets_discovery: input.maxWalletsDiscovery ?? 100,
      max_runtime_seconds: input.maxRuntimeSeconds ?? 12,
    }),
    method: "POST",
  }, 30000);
}

export async function runWalletAnalysisJobStep(input: {
  batchSize?: number;
  historyLimit?: number;
  jobId: string;
  maxWallets?: number;
  maxWalletsDiscovery?: number;
  maxRuntimeSeconds?: number;
}): Promise<WalletAnalysisRunResponse> {
  return fetchApiJson<WalletAnalysisRunResponse>(`/wallet-analysis/jobs/${encodeURIComponent(input.jobId)}/run-step`, {
    body: JSON.stringify({
      batch_size: input.batchSize ?? 10,
      history_limit: input.historyLimit ?? 100,
      max_wallets: input.maxWallets ?? 50,
      max_wallets_discovery: input.maxWalletsDiscovery ?? 100,
      max_runtime_seconds: input.maxRuntimeSeconds ?? 12,
    }),
    method: "POST",
  }, 20000);
}

export async function fetchWalletAnalysisJob(jobId: string): Promise<WalletAnalysisJobRead> {
  return fetchApiJson<WalletAnalysisJobRead>(`/wallet-analysis/jobs/${encodeURIComponent(jobId)}`);
}

export async function fetchWalletAnalysisCandidates(input: {
  confidence?: WalletAnalysisConfidence;
  jobId: string;
  limit?: number;
  offset?: number;
  outcome?: string;
  side?: string;
  sortBy?: WalletAnalysisCandidateSortBy;
  sortOrder?: WalletAnalysisSortOrder;
}): Promise<WalletAnalysisCandidateList> {
  const params = new URLSearchParams();
  if (input.side) {
    params.set("side", input.side);
  }
  if (input.outcome) {
    params.set("outcome", input.outcome);
  }
  if (input.confidence) {
    params.set("confidence", input.confidence);
  }
  params.set("sort_by", input.sortBy ?? "score");
  params.set("sort_order", input.sortOrder ?? "desc");
  params.set("limit", String(input.limit ?? 10));
  params.set("offset", String(input.offset ?? 0));
  return fetchApiJson<WalletAnalysisCandidateList>(
    `/wallet-analysis/jobs/${encodeURIComponent(input.jobId)}/candidates?${params.toString()}`,
  );
}

export async function saveWalletAnalysisCandidateAsProfile(candidateId: string): Promise<WalletProfileRead> {
  return fetchApiJson<WalletProfileRead>(
    `/wallet-analysis/candidates/${encodeURIComponent(candidateId)}/save-profile`,
    {
      body: "{}",
      method: "POST",
    },
  );
}

export async function fetchWalletProfiles(input?: {
  status?: WalletProfileStatus;
  limit?: number;
  offset?: number;
}): Promise<WalletProfileList> {
  const params = new URLSearchParams();
  params.set("limit", String(input?.limit ?? 20));
  params.set("offset", String(input?.offset ?? 0));
  if (input?.status) {
    params.set("status", input.status);
  }
  return fetchApiJson<WalletProfileList>(`/wallet-profiles?${params.toString()}`);
}

export async function updateWalletProfile(
  profileId: string,
  payload: {
    alias?: string | null;
    status?: WalletProfileStatus;
    notes?: string | null;
  },
): Promise<WalletProfileRead> {
  return fetchApiJson<WalletProfileRead>(`/wallet-profiles/${encodeURIComponent(profileId)}`, {
    body: JSON.stringify(payload),
    method: "PATCH",
  });
}

export async function followWalletProfileInDemo(profileId: string): Promise<WalletProfileDemoFollowResponse> {
  return fetchApiJson<WalletProfileDemoFollowResponse>(
    `/wallet-profiles/${encodeURIComponent(profileId)}/demo-follow`,
    {
      body: "{}",
      method: "POST",
    },
  );
}

export async function fetchPolySignalMarketSignals(input?: {
  confidence?: WalletAnalysisConfidence;
  jobId?: string;
  limit?: number;
  marketSlug?: string;
  offset?: number;
  predictedSide?: string;
  signalStatus?: MarketSignalStatus;
}): Promise<PolySignalMarketSignalList> {
  const params = new URLSearchParams();
  params.set("limit", String(input?.limit ?? 10));
  params.set("offset", String(input?.offset ?? 0));
  if (input?.jobId) {
    params.set("job_id", input.jobId);
  }
  if (input?.marketSlug) {
    params.set("market_slug", input.marketSlug);
  }
  if (input?.predictedSide) {
    params.set("predicted_side", input.predictedSide);
  }
  if (input?.confidence) {
    params.set("confidence", input.confidence);
  }
  if (input?.signalStatus) {
    params.set("signal_status", input.signalStatus);
  }
  return fetchApiJson<PolySignalMarketSignalList>(`/polysignal-market-signals?${params.toString()}`);
}

export async function settlePolySignalMarketSignal(signalId: string): Promise<PolySignalMarketSignalSettlement> {
  return fetchApiJson<PolySignalMarketSignalSettlement>(
    `/polysignal-market-signals/${encodeURIComponent(signalId)}/settle`,
    {
      body: "{}",
      method: "POST",
    },
  );
}

export async function settlePendingPolySignalMarketSignals(input?: {
  jobId?: string;
  limit?: number;
  marketSlug?: string;
}): Promise<PolySignalMarketSignalSettlePendingResponse> {
  return fetchApiJson<PolySignalMarketSignalSettlePendingResponse>(
    "/polysignal-market-signals/settle-pending",
    {
      body: JSON.stringify({
        job_id: input?.jobId,
        limit: input?.limit ?? 10,
        market_slug: input?.marketSlug,
      }),
      method: "POST",
    },
  );
}

File: apps/api/app/schemas/wallet_analysis.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal, InvalidOperation
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

WalletProfileStatus = Literal["candidate", "watching", "demo_follow", "paused", "rejected"]
WalletMetricStatus = Literal["verified", "estimated", "unavailable"]
WalletConfidence = Literal["low", "medium", "high"]
WalletSignalStrength = Literal["weak", "moderate", "strong"]
WalletAnalysisJobStatus = Literal[
    "pending",
    "resolving_market",
    "discovering_wallets",
    "analyzing_wallets",
    "scoring",
    "completed",
    "partial",
    "failed",
    "cancelled",
]
WalletAnalysisCandidateSortBy = Literal["score", "volume_30d", "win_rate_30d", "pnl_30d", "created_at"]
SortOrder = Literal["asc", "desc"]
WalletAnalysisResolveStatus = Literal["ok", "partial", "not_found", "unsupported", "error"]
WalletAnalysisRunState = Literal["progressed", "already_running", "no_work_remaining", "failed"]
MarketSignalStatus = Literal[
    "pending_resolution",
    "resolved_hit",
    "resolved_miss",
    "cancelled",
    "unknown",
    "no_clear_signal",
]
MarketResolutionStatus = Literal["open", "resolved", "cancelled", "unknown"]


class WalletAnalysisCreateRequest(BaseModel):
    polymarket_url: str = Field(max_length=2048)


class WalletAnalysisOutcomeRead(BaseModel):
    label: str
    side: str
    token_id: str | None = None


class WalletAnalysisResolvedLinkRead(BaseModel):
    source_url: str
    normalized_url: str
    status: WalletAnalysisResolveStatus
    raw_source: str
    market_title: str | None = None
    condition_id: str | None = None
    market_slug: str | None = None
    event_slug: str | None = None
    sport_or_league: str | None = None
    outcomes: list[WalletAnalysisOutcomeRead] = Field(default_factory=list)
    token_ids: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class WalletAnalysisJobProgressRead(BaseModel):
    wallets_found: int = 0
    wallets_analyzed: int = 0
    wallets_with_sufficient_history: int = 0
    yes_wallets: int = 0
    no_wallets: int = 0
    current_batch: int = 0


class WalletAnalysisSignalSummaryRead(BaseModel):
    id: str
    predicted_side: str | None = None
    predicted_outcome: str | None = None
    polysignal_score: Decimal | None = None
    confidence: WalletConfidence
    data_confidence: WalletConfidence | None = None
    signal_strength: WalletSignalStrength | None = None
    signal_margin: Decimal | None = None
    yes_score: Decimal | None = None
    no_score: Decimal | None = None
    outcome_scores_json: dict[str, Any] | None = None
    outcome_wallet_counts_json: dict[str, int] | None = None
    signal_status: MarketSignalStatus
    warnings_json: list[str] = Field(default_factory=list)


class WalletAnalysisJobRead(BaseModel):
    id: str
    source_url: str
    normalized_url: str
    market_slug: str | None = None
    event_slug: str | None = None
    condition_id: str | None = None
    market_title: str | None = None
    status: WalletAnalysisJobStatus
    outcomes: list[WalletAnalysisOutcomeRead] = Field(default_factory=list)
    token_ids: list[str] = Field(default_factory=list)
    progress: WalletAnalysisJobProgressRead
    result_json: dict[str, Any] | None = None
    warnings: list[str] = Field(default_factory=list)
    status_detail: str | None = None
    error_message: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    candidates_count: int = 0
    signal_summary: WalletAnalysisSignalSummaryRead | None = None


class WalletAnalysisJobCreateResponse(BaseModel):
    job_id: str
    status: WalletAnalysisJobStatus
    message: str
    market: WalletAnalysisJobRead


class WalletAnalysisJobRunRequest(BaseModel):
    max_wallets: int = Field(default=50, ge=1, le=250)
    max_wallets_discovery: int = Field(default=100, ge=1, le=500)
    batch_size: int = Field(default=20, ge=1, le=100)
    history_limit: int = Field(default=100, ge=1, le=250)
    max_runtime_seconds: int = Field(default=12, ge=5, le=25)


class WalletAnalysisJobRunResponse(BaseModel):
    job_id: str
    status: WalletAnalysisJobStatus
    run_state: WalletAnalysisRunState
    message: str
    wallets_found: int
    wallets_analyzed: int
    wallets_with_sufficient_history: int
    candidates_count: int
    warnings: list[str] = Field(default_factory=list)
    status_detail: str | None = None
    has_more: bool = False
    next_action: str | None = None
    signal_id: str | None = None
    signal_status: MarketSignalStatus | None = None
    market: WalletAnalysisJobRead


class WalletAnalysisCandidateRead(BaseModel):
    id: str
    job_id: str
    wallet_address: str
    outcome: str | None = None
    side: str | None = None
    token_id: str | None = None
    observed_market_position_usd: Decimal | None = None
    score: Decimal | None = None
    confidence: WalletConfidence
    roi_30d_status: WalletMetricStatus
    roi_30d_value: Decimal | None = None
    win_rate_30d_status: WalletMetricStatus
    win_rate_30d_value: Decimal | None = None
    pnl_30d_status: WalletMetricStatus
    pnl_30d_value: Decimal | None = None
    trades_30d: int | None = None
    volume_30d: Decimal | None = None
    markets_traded_30d: int | None = None
    last_activity_at: datetime | None = None
    reasons_json: list[str] = Field(default_factory=list)
    risks_json: list[str] = Field(default_factory=list)
    raw_summary_json: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime


class WalletAnalysisCandidateList(BaseModel):
    items: list[WalletAnalysisCandidateRead]
    total: int


class WalletProfileUpsert(BaseModel):
    wallet_address: str = Field(max_length=42)
    alias: str | None = Field(default=None, max_length=160)
    status: WalletProfileStatus = "candidate"
    score: Decimal | None = None
    confidence: WalletConfidence = "low"
    roi_30d_status: WalletMetricStatus = "unavailable"
    roi_30d_value: Decimal | None = None
    win_rate_30d_status: WalletMetricStatus = "unavailable"
    win_rate_30d_value: Decimal | None = None
    pnl_30d_status: WalletMetricStatus = "unavailable"
    pnl_30d_value: Decimal | None = None
    trades_30d: int | None = Field(default=None, ge=0)
    volume_30d: Decimal | None = None
    drawdown_30d_status: WalletMetricStatus = "unavailable"
    drawdown_30d_value: Decimal | None = None
    markets_traded_30d: int | None = Field(default=None, ge=0)
    last_activity_at: datetime | None = None
    discovered_from_market: str | None = Field(default=None, max_length=320)
    discovered_from_url: str | None = Field(default=None, max_length=512)
    discovered_at: datetime | None = None
    reasons_json: list[str] = Field(default_factory=list, max_length=50)
    risks_json: list[str] = Field(default_factory=list, max_length=50)
    notes: str | None = Field(default=None, max_length=4000)

    model_config = ConfigDict(populate_by_name=True)

    @field_validator("wallet_address", mode="before")
    @classmethod
    def normalize_wallet_address(cls, value: object) -> str:
        if not isinstance(value, str):
            raise ValueError("wallet_address must be a public wallet address")
        return value.strip().lower()

    @field_validator(
        "score",
        "roi_30d_value",
        "win_rate_30d_value",
        "pnl_30d_value",
        "volume_30d",
        "drawdown_30d_value",
        mode="before",
    )
    @classmethod
    def normalize_decimal(cls, value: object) -> Decimal | None:
        if value is None:
            return None
        if isinstance(value, Decimal):
            return value
        if isinstance(value, int | float | str):
            try:
                return Decimal(str(value).strip())
            except (InvalidOperation, ValueError):
                return None
        return None


class WalletProfileRead(BaseModel):
    id: str
    wallet_address: str
    alias: str | None = None
    status: WalletProfileStatus
    score: Decimal | None = None
    confidence: WalletConfidence
    roi_30d_status: WalletMetricStatus
    roi_30d_value: Decimal | None = None
    win_rate_30d_status: WalletMetricStatus
    win_rate_30d_value: Decimal | None = None
    pnl_30d_status: WalletMetricStatus
    pnl_30d_value: Decimal | None = None
    trades_30d: int | None = None
    volume_30d: Decimal | None = None
    drawdown_30d_status: WalletMetricStatus
    drawdown_30d_value: Decimal | None = None
    markets_traded_30d: int | None = None
    last_activity_at: datetime | None = None
    discovered_from_market: str | None = None
    discovered_from_url: str | None = None
    discovered_at: datetime | None = None
    reasons_json: list[str] = Field(default_factory=list)
    risks_json: list[str] = Field(default_factory=list)
    notes: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WalletProfileList(BaseModel):
    items: list[WalletProfileRead]
    total: int


class WalletProfileUpdate(BaseModel):
    alias: str | None = Field(default=None, max_length=160)
    status: WalletProfileStatus | None = None
    notes: str | None = Field(default=None, max_length=4000)


class WalletProfileDemoFollowResponse(BaseModel):
    profile: WalletProfileRead
    copy_wallet: dict[str, Any]
    already_following: bool = False
    baseline_created_at: datetime
    message: str


class PolySignalMarketSignalRead(BaseModel):
    id: str
    job_id: str | None = None
    source_url: str | None = None
    market_slug: str | None = None
    event_slug: str | None = None
    condition_id: str | None = None
    market_title: str | None = None
    outcomes_json: list[dict[str, Any]] = Field(default_factory=list)
    token_ids_json: list[str] = Field(default_factory=list)
    predicted_side: str | None = None
    predicted_outcome: str | None = None
    polysignal_score: Decimal | None = None
    confidence: WalletConfidence
    data_confidence: WalletConfidence | None = None
    signal_strength: WalletSignalStrength | None = None
    signal_margin: Decimal | None = None
    yes_score: Decimal | None = None
    no_score: Decimal | None = None
    outcome_scores_json: dict[str, Any] | None = None
    outcome_wallet_counts_json: dict[str, int] | None = None
    wallets_analyzed: int | None = None
    wallets_with_sufficient_history: int | None = None
    top_wallets_json: list[dict[str, Any]] = Field(default_factory=list)
    warnings_json: list[str] = Field(default_factory=list)
    signal_status: MarketSignalStatus
    final_outcome: str | None = None
    final_resolution_source: str | None = None
    resolved_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class PolySignalMarketResolutionRead(BaseModel):
    status: MarketResolutionStatus
    final_outcome: str | None = None
    source: str
    confidence: WalletConfidence
    reason: str
    checked_at: datetime


class PolySignalMarketSignalSettlementRead(BaseModel):
    signal: PolySignalMarketSignalRead
    resolution: PolySignalMarketResolutionRead
    changed: bool = False


class PolySignalMarketSignalMetricsBucketRead(BaseModel):
    total: int = 0
    resolved_hit: int = 0
    resolved_miss: int = 0
    win_rate: Decimal | None = None


class PolySignalMarketSignalMetricsRead(BaseModel):
    total: int = 0
    pending_resolution: int = 0
    resolved_hit: int = 0
    resolved_miss: int = 0
    cancelled: int = 0
    unknown: int = 0
    no_clear_signal: int = 0
    win_rate: Decimal | None = None
    avg_score_resolved_hit: Decimal | None = None
    avg_score_resolved_miss: Decimal | None = None
    by_confidence: dict[str, PolySignalMarketSignalMetricsBucketRead] = Field(default_factory=dict)


class PolySignalMarketSignalList(BaseModel):
    items: list[PolySignalMarketSignalRead]
    total: int
    metrics: PolySignalMarketSignalMetricsRead


class PolySignalMarketSignalSettlePendingRequest(BaseModel):
    limit: int = Field(default=10, ge=1, le=50)
    job_id: str | None = None
    market_slug: str | None = Field(default=None, max_length=256)


class PolySignalMarketSignalSettlePendingResponse(BaseModel):
    checked: int = 0
    still_pending: int = 0
    resolved_hit: int = 0
    resolved_miss: int = 0
    cancelled: int = 0
    unknown: int = 0
    errors: int = 0
    items: list[PolySignalMarketSignalSettlementRead] = Field(default_factory=list)
